### ISSUE TITLE
Export to hml file with self-contained images

### ISSUE THREAD

**Issue by aloboa** [zim-desktop-wiki/zim-desktop-wiki#2175]
I would like to be able to export to a self-contained, standalone HTML file, including images, like, eg. from RStudio:
![image](https://user-images.githubusercontent.com/6074054/194339853-d057f104-fc96-405a-ab24-8ca29f5977ad.png)

**Comment by jaap-karssenberg** [comment]
Have you tried export to "web archive" ? THis creates an ".mht" which is
self contained html package.

I'm not aware of a different standard for really self contained files.

--- Jaap

**Comment by bruvellu** [comment]
Pandoc has an option to generate a self-contained HTML file using the [data URI scheme](https://en.wikipedia.org/wiki/Data_URI_scheme):

`--embed-resources --standalone` (formerly `--self-contained`).

See https://pandoc.org/MANUAL.html#options-affecting-specific-writers

**Comment by menthalo37** [comment]
.mht export (archive) is a good solution. Even if such format seems to be desuet.

**Comment by aloboa** [comment]
> I'm not aware of a different standard for really self contained files.

Surprising, as this is available in RStudio since years ago:
File created with ZIm: https://www.dropbox.com/s/7hwksyekeu6ghro/Test.txt?dl=0
md exported from zim: https://www.dropbox.com/s/f2zaphzt9meq7rl/Test.md?dl=0
(+ a folder with the 2 images)
HTML created by RStudio (without any additional folder): https://diapiro.geo3bcn.csic.es/alobo/blog/Test.html

> .mht export (archive) is a good solution. Even if such format seems to be desuet.

No, it is not a good solution. I asked for an HTML file in this thread, not any other format.

**Comment by bruvellu** [comment]
Indeed RStudio's RMarkdown uses Pandoc to render HTML (see [recent related issue/PR](https://github.com/rstudio/rmarkdown/issues/2382)). But Pandoc cannot (yet) convert _from_ Zim Wiki Syntax to other formats. Perhaps Zim's regular HTML export can be optionally piped through Pandoc to get a self-contained file, otherwise this will require a new [custom export format](https://github.com/zim-desktop-wiki/zim-desktop-wiki/tree/develop/zim/formats). Have a look at the code and feel free to make a pull request suggesting an implementation.

For now the simplest workaround is to:

1. Export as HTML from Zim
2. Run `pandoc Exported.html --embed-resources --standalone --output=Contained.html`

Ps: For Pandoc versions earlier than v2.19.2 replace `--embed-resources --standalone` with `--self-contained`.

**Comment by aloboa** [comment]
@nelas. Thanks for your useful answer. Nevertheles,
1. If the workaround you suggest is so simple, why cannot it be included within the regular zim menu instead of requiring the user to run an additional command outside Zim?
2. Where you say

> Have a look at the code and feel free to make a pull request suggesting an implementation.

Do not really know how to do it. Perhaps you can do such a pull request?

**Comment by introt** [comment]
A `pandoc` installation is larger than Zim itself, so it doesn't make sense to bundle it. Someone could develop a plugin to streamline it though - or the user could add the command as a custom tool.

It's also possible to base64 encode the image files and stash them in `<img>` tags, similar to how our checkboxes are currently baked into the export template. I could try look into this around 2030 given my current backlog :)

**Comment by introt** [comment]
@aloboa Here's a plugin that makes the HTML exporter inline images. Go to Preferences > Plugins > Open plugins folder and place the file there, then restart Zim and enable the plugin. Merry Christmas || Happy Holidays!

https://gist.github.com/introt/fed672dfbea9df399610084be448326b

**Comment by aloboa** [comment]
@introt
Thanks, 
Happy new year!
Unfortunately,
I have copied the *.py file to the plugins folder
/home/alobo/.var/app/org.zim_wiki.Zim/data/zim/plugins
but get an error
"There was a problem loading this plugin"
when I select it from the list in Preferences/Plugins

Using 0.75.1

**Comment by introt** [comment]
@aloboa Let's see what can be done about that over at the gist - I'll tag you there

**Comment by aloboa** [comment]
The plugin works with 0.75.2, but export still creates a folder. If I move the HTML only, the images are there
but the style is lost:
![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/6074054/43d5d011-32ce-494e-b004-06a37a5d7fca)



Is there any way the style information could be also embedded in the HTML file, so that the HTML would be completely self-contained?

HTML in the same directory as the folder:
![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/6074054/3de6c2c4-b777-4795-9107-32e4dd3b82cf)


HTML moved to another directory:
![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/6074054/b136b32d-af25-44bc-8756-eb1262c417cb)

**Comment by introt** [comment]
> Is there any way the style information could be also embedded in the HTML file, so that the HTML would be completely self-contained?

Yes, and so can the SVG icons seen in _resources. I'll put this on my backlog, but in the meantime you can open the HTML file, remove the references to the CSS style files and instead use a `<style>` tag. See "Internal CSS" here: https://www.w3schools.com/html/html_css.asp

Iincluding the SVGs is quite similar, just replace the references inside the `<svg>` tags with the code contained in the SVG files: https://www.w3schools.com/graphics/svg_inhtml.asp

**Comment by aloboa** [comment]
@introt Any chance you could find time for this improvement any time soon?

**Comment by introt** [comment]
I'm afraid I can't give you an ETA on that, I'm currently quite tied up with my degree
